### ISSUE TITLE
Bug with LCP profiles parsing

### ISSUE THREAD

**Issue by kznsq** [Beep6581/RawTherapee#3788]
The XML parser do not accumulate the text while parsing with the 'LCPProfile::XmlTextHandler' and causes wrong reading numbers.
You can set the small (3 bytes, for example) BufferSize in the LCPProfile::LCPProfile and debug values readed, for example, here:
LCPProfile::XmlTextHandler(..)
{
..
pProf->pCurCommon->imgXCenter = atof(raw);
..
}

**Comment by agriggio** [comment]
Fix in 633af0b (branch `lcp-parser-fix-xml-handle-text`). As usual, I went for the smallest patch (the LCP code would benefit from a general cleanup, but that's a separate issue). Please test, thanks!

**Comment by agriggio** [comment]
@kznsq do you have an LCP profile that used to make the parser fail? I'd like to close this but I want to test your original scenario first. Thanks!

**Comment by kznsq** [comment]
Hi!
Here the file.

Thanks for your response!

Also the LCPMapper needs to support fisheye mapping like this

void LCPMapper::correctDistortion(float& x, float& y) const
{
    if(!isFisheye) // flag needs to be added
    {
        double xd = (x - mc.x0) / mc.fx, yd = (y - mc.y0) / mc.fy;

        const float* aDist = mc.param;
        double rsqr = xd * xd + yd * yd;
        double xfac = aDist[swapXY ? 3 : 4], yfac = aDist[swapXY ? 4 : 3];

        double commonFac = (((aDist[2] * rsqr + aDist[1]) * rsqr +
aDist[0]) * rsqr + 1.)
                           + 2. * (yfac * yd + xfac * xd);

        double xnew = xd * commonFac + xfac * rsqr;
        double ynew = yd * commonFac + yfac * rsqr;

        x = xnew * mc.fx + mc.x0;
        y = ynew * mc.fy + mc.y0;
    }
    else
    {
        double u = x;
        double v = y;
        double u0 = mc.x0;
        double v0 = mc.y0;
        double du = u - u0;
        double dv = v - v0;
        double fx = mc.fx;
        double fy = mc.fy;
        double k1 = mc.param[0];
        double k2 = mc.param[1];
        double r = std::sqrt(du * du + dv * dv);
        double f = std::sqrt(fx*fy);
        double th = std::atan(r/f);
        double thCommon = (1.0/r)*(th + k1*th*th*th + k2*th*th*th*th*th);
        float ud = thCommon*fx*du + u0;
        float vd = thCommon*fy*dv + v0;

        x = ud;
        y = vd;
    }
}

**Comment by agriggio** [comment]
thanks for the quick reply @kznsq , but I don't see a file attached, sorry

**Comment by kznsq** [comment]
Sorry, maybe email server cuts off the attached file?
This file is "NIKON D3X (Nikon FX 14-24mm f2.8 G ED) - RAW.lcp" from the
latest Adobe Camera Raw

**Comment by agriggio** [comment]
@kznsq thanks. Do you happen to have also a RAW file with which I can test?

**Comment by agriggio** [comment]
@kznsq nevermind, I was able to test without a RAW. The fix works, I'll merge asap. Thanks for the help!

**Comment by kznsq** [comment]
Thank you )

**Comment by agriggio** [comment]
the fisheye distortion problem is however a different issue. For that, I would definitely need a RAW shot with a fisheye lens for which there is a LCP profile. I don't have any, maybe you could help?

**Comment by kznsq** [comment]
You can try this one:
https://s3.amazonaws.com/Sample_NEFs/fisheye.nef.zip

**Comment by Beep6581** [comment]
> maybe email server cuts off the attached file?

@kznsq GitHub does not allow LCP attachments, only PNG, GIF, JPG, DOCX, PPTX, XLSX, TXT, PDF, and ZIP. In the future, please upload using https://filebin.net/

**Comment by Beep6581** [comment]
@agriggio I can't promise that this LCP is compatible with the lens I used when taking the photo, but by the name of it it sounds right, so maybe it helps you test: https://filebin.net/f4xmixo8wmzdsj9o

**Comment by agriggio** [comment]
@kznsq can you clarify in which sense the current distortion correction for fisheye is not working? I just tried with lensfun (via both gimp-lensfun and ufraw) and I get the exact same result as RT.

**Comment by agriggio** [comment]
parser fix merged into dev

**Comment by Beep6581** [comment]
Does the patch in [this comment](https://github.com/Beep6581/RawTherapee/issues/3788#issuecomment-292490483) let the LCP reproject the image from fisheye to rectilinear?

**Comment by kznsq** [comment]
Maybe I didn't  mention something... but there are 2 models for distortion in the Adobe's document on LCP files: Rectilinear lens and Fisheye  
http://download.macromedia.com/pub/labs/lensprofile_creator/lensprofile_creator_cameramodel.pdf

**Comment by kznsq** [comment]
> Does the patch in this comment let the LCP reproject the image from fisheye to rectilinear?

Yes, I hope so :)

**Comment by Beep6581** [comment]
Would be great if RT supported de-fishing!

**Comment by agriggio** [comment]
So, should RT select automatically the fisheye model for fisheye lenses? This doesn't seem to match what LensFun-powered tools do (e.g. the GIMP lensfun plugin or ufraw). If this is the desired behaviour, then I think the patch goes in the right direction. It seems to be still a bit inaccurate, at least when compared with ufraw when explicitly selecting a rectilinear target geometry. See the two attached images.

RT with @kznsq patch:

![fisheye-rt-kznsq](https://cloud.githubusercontent.com/assets/11982086/24804893/37839cb2-1bb0-11e7-9c5f-2d58a39a4af2.jpg)

ufraw with rectilinear target geometry (and auto-cropping):

![fisheye-ufraw-rectilinear](https://cloud.githubusercontent.com/assets/11982086/24804933/526f809a-1bb0-11e7-83c6-959e9311452c.jpg)

**Comment by Beep6581** [comment]
> should RT select automatically the fisheye model for fisheye lenses?

That's what I would expect as a user.

Does your UFRaw example use the same LCP, or does it use lensfun?
Could you show a comparison using my sample photo+LCP?

**Comment by agriggio** [comment]
> That's what I would expect as a user.

Ok, great. I've never used a fisheye so I don't really know what these users expect :smile: 
I'll try to see how to fix the correction then (we agree that it's still not precise, don't we?)

**Comment by kznsq** [comment]
Maybe LCP is not correct enough as lensfun in some cases?
These patches can detec the fisheye:

class LCPMapper
{
...
    bool isFisheye;
...
};
 
// if !vignette then geometric and CA
LCPMapper::LCPMapper(LCPProfile* pProf, float focalLength, float focalLength35mm, float focusDist, float aperture, bool vignette, bool useCADistP,
                     int fullWidth, int fullHeight, const CoarseTransformParams& coarse, int rawRotationDeg)
{
...    
    isFisheye = pProf->isFisheye;
}

**Comment by agriggio** [comment]
> Maybe LCP is not correct enough as lensfun in some cases?

I don't know yet -- I have to investigate :-)

**Comment by kznsq** [comment]
<img width="1440" alt="screen shot 2017-04-07 at 5 54 55 pm" src="https://cloud.githubusercontent.com/assets/5992301/24805730/76299150-1bbb-11e7-8c55-2620960a3f63.png">
The result in the LR seems better than the result after my patch, maybe it needs to be improved...

**Comment by Beep6581** [comment]
Not knowing how the LCP code works, I wanted to point out that there are various fisheye projections, and that reprojecting an image (from fisheye to rectilinear) is not the same as correcting distortions. It could be the case that RT should not only reproject, but also correct distortions (before or after reprojecting).

**Comment by agriggio** [comment]
Just pushed a `lcp-fisheye` branch that applies the correct "distortion correction" for fisheye. As @Beep6581 said already, this is really a distortion correction + change of projection, but the Adobe manual simple calls it "distortion correction", and this is how it appears in the RT gui as well.

The patch of @kznsq was applying the Adobe specs correctly, but the issue was that the order of transformations applied by RT was not correct. The code was correcting distortion first, and then scaling (for auto-fill) next. But the right order is scale first, and correct distortion next. See e.g. http://lensfun.sourceforge.net/manual/corrections.html

Please test, thanks!

**Comment by kznsq** [comment]
Here is the my corrected image. Looks like the LR result.

<img width="1047" alt="screen shot 2017-04-08 at 12 01 27 pm" src="https://cloud.githubusercontent.com/assets/5992301/24827496/699991ea-1c53-11e7-9dc8-ba0bbe517b86.png">

**Comment by kznsq** [comment]
Scale was applied after the distortion, in reverse order.

**Comment by Beep6581** [comment]
Before: https://i.imgur.com/ooezLJd.jpg
After: https://i.imgur.com/8nKE3eM.jpg
Woo! Thank you.

**Comment by agriggio** [comment]
@Beep6581 great :smile: BTW, make sure you have "auto-fill" turned on otherwise you get an insane amount of cropping... I still have to address that, but I think it's a separate issue

**Comment by kznsq** [comment]
Great! :)

**Comment by heckflosse** [comment]
:+1:
